server/controllers/userControllers/rentals.js

```diff
@@ -69,3 +97,111 @@ exports.updateRentalPaidStatus = async (req, res) => {
+    }
+
+    const userDetails = await User.findById(rentalDetails.user);
+
+    const start = new Date(newStartDate);
+    const end = new Date(newEndDate);
+    
+    if(rentalDetails.paid === true){
+      return res.status(400).json({ message: "Cannot update dates for a paid rental." });
+    }
+
+    if (start >= end) {
+      return res.status(400).json({ message: "End date should be greater than start date." });
+    }
+
+    const overlappingRentals = carDetails.rentals.filter((rental) => {
+      return start <= rental.endDate && end >= rental.startDate && rental.rentalId.toString() !== rentalId;
+    });
+
+    if (overlappingRentals.length >= carDetails.quantity) {
+      return res.status(400).json({
+        message: "Car is already rented for the requested dates."
+      });
+    }
+
+    const pricePerDay = carDetails.pricePerDay;
+    const diffTime = Math.abs(end - start);
+    const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24)) + 2; 
+    const newPrice = diffDays * pricePerDay;
+
+    rentalDetails.startDate = start;
+    rentalDetails.endDate = end;
+    rentalDetails.price = newPrice;
+    await rentalDetails.save();
+
+    carDetails.rentals = carDetails.rentals.filter(rental => rental.rentalId.toString() !== rentalDetails._id.toString());
+    carDetails.rentals.push({
+      rentalId: rentalId,
+      startDate: start,
+      endDate: end
+    });
+    await carDetails.save();
+
+    userDetails.rentals = userDetails.rentals.filter(rental => rental.rentalId.toString() !== rentalDetails._id.toString());
+    userDetails.rentals.push({
+      rentalId: rentalId,
+      startDate: start,
+      endDate: end,
+      price: newPrice
+    });
+    await userDetails.save();
+
+    res.status(200).json({
+      message: "Rental updated successfully",
+      newStartDate: start,
+      newEndDate: end,
+      newPrice: newPrice
+    });
+  } catch (error) {
+    console.error(error);
+    res.status(500).json({ error: "Internal server error" });
+  }
+};
+
+exports.deleteRemoveRental = async (req, res) => {
+
+  const params = req.params;
+  const rentalId =  params.id;
+
+  try {
+    const rentalDetails = await Rental.findById(rentalId);
+    if (!rentalDetails) {
+      return res.status(404).json({
+        message: "Rental not found"
+      });
+    }
+
+    const carDetails = await Car.findById(rentalDetails.car);
+    const userDetails = await User.findById(rentalDetails.user);
+  
+    carDetails.rentals = carDetails.rentals.filter(rental => rental.rentalId.toString() !== rentalId);
+    await carDetails.save();
+
+    userDetails.rentals = userDetails.rentals.filter(rental => rental.rentalId.toString() !== rentalId);
+    await userDetails.save();
+
+    await Rental.deleteOne({ _id: rentalId });
+
+    res.status(200).json({ message: "Rental removed successfully" });
+  } catch (error) {
+    console.error(error);
+    res.status(500).json({ error: "Internal server error" });
+  }
+};
```

Comment:
Mamy juz takiego route'a dla "admina", więc nie ma sensu mieć dwa razy tego samego chyba. Jeżeli coś może user i admin to zostawmy to jako user route.